opentreemap/treemap/udf.py

```diff
@@ -990,132 +1202,154 @@ def __getitem__(self, key):
+            self.instance.dirty_collection_udfs.add(key)
+            # Don't let the caller update collection-subfields on the fly.
+            # They have to make their change, and then set the value here
+            # again, in order to go through this entrypoint.
             self.collection_fields[key] = copy.deepcopy(val)
         else:
-            val = udf.reverse_clean(val)
             super(UDFDictionary, self).__setitem__(key, val)
 
+    def __delitem__(self, key):
+        udf = self._get_udf_or_error(key)
 
-class UDFDescriptor(Creator):
-    def __get__(self, obj, type=None):
-        if obj is None:
-            return None
-        # UDFDictionary needs a reference to the model instance to lookup
-        # collection UDFs
-        udf_dict = obj.__dict__[self.field.name]
-        # Workaround for test failure on some dev machines
-        if udf_dict == '':
-            udf_dict = UDFDictionary({}, self.field, obj)
-        udf_dict.instance = obj
-
-        return udf_dict
-
-    def __set__(self, obj, value):
-        value = self.field.to_python(value)
-        if isinstance(value, dict):
-            value = UDFDictionary(
-                value=value, field=self.field, instance=obj
-            )
-        # setattr goes into infinite recursion, so fish in `__dict__`
-        obj.__dict__[self.field.name] = value
-
-
-class UDFField(DictionaryField):
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def get_default(self):
-        hstore_dict = super(UDFField, self).get_default()
-        if isinstance(hstore_dict, HStoreDict):
-            return UDFDictionary(hstore_dict, self)
+        if udf.iscollection:
+            self.instance.dirty_collection_udfs.add(key)
+            del self.collection_fields[key]
         else:
-            return hstore_dict
+            super(UDFDictionary, self).__delitem__(key)
 
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def get_prep_value(self, value):
-        if isinstance(value, dict) and not isinstance(value, UDFDictionary):
-            return UDFDictionary(value, self)
-        else:
-            return value
+    def clean_value(self):
+        '''
+        Returns a dict whose keys are the same as self,
+        and whose values are converted to either string or None,
+        as expected by the db.
+        '''
+        udfds = {udfd.name: udfd for udfd in
+                 self.instance.get_user_defined_fields()
+                 if not udfd.iscollection}
 
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def contribute_to_class(self, cls, name):
-        super(UDFField, self).contribute_to_class(cls, name)
-        setattr(cls, self.name, UDFDescriptor(self))
+        return {key: udfds[key].reverse_clean(val)
+                for (key, val) in super(UDFDictionary, self).iteritems()}
 
+    # The automated tests require filling in all the collection methods
```

Comment:
I think `iteritems` is still needed in order to iterate through collectcion udfs. But I was able to delete the rest and have the tests pass.